ceno_zkvm/src/instructions/riscv/div.rs

```diff
@@ -128,43 +352,122 @@ impl<E: ExtensionField, I: RIVInstruction> Instruction<E> for ArithInstruction<E
+
+                is_dividend_max_negative.assign_instance(
+                    instance,
+                    (dividend as u32 as u64).into(),
+                    (i32::MIN as u32 as u64).into(),
+                )?;
+                is_divisor_minus_one.assign_instance(
+                    instance,
+                    (divisor as u32 as u64).into(),
+                    (-1i32 as u32 as u64).into(),
+                )?;
+
+                let signed_div_overflow_b = dividend == i32::MIN && divisor == -1i32;
+                set_val!(instance, is_signed_overflow, signed_div_overflow_b as u64);
+
+                quotient_signed.assign_instance(instance, lkm, &quotient_v)?;
+                remainder_signed.assign_instance(instance, lkm, &remainder_v)?;
+
+                let negate_if = |b: bool, x: i32| if b { -(x as i64) } else { x as i64 };
+
+                let remainder_pos_orientation = negate_if(negative_division_b, remainder);
+                let divisor_pos_orientation = negate_if(divisor < 0, divisor);
+
+                remainder_nonnegative.assign_instance(
+                    instance,
+                    lkm,
+                    (-Goldilocks::ONE).to_canonical_u64(),
+                    i64_to_base::<Goldilocks>(remainder_pos_orientation).to_canonical_u64(),
+                )?;
+
+                (
+                    remainder_pos_orientation as u32,
+                    divisor_pos_orientation as u32,
+                )
+            }
+        };
 
-        // assignment
-        config.r_insn.assign_instance(instance, lkm, step)?;
-        config
-            .divisor
-            .assign_limbs(instance, divisor.as_u16_limbs());
-        config
-            .outcome
-            .assign_limbs(instance, outcome.as_u16_limbs());
+        config.dividend.assign_value(instance, dividend_v);
+        config.divisor.assign_value(instance, divisor_v);
+        config.quotient.assign_value(instance, quotient_v);
+        config.remainder.assign_value(instance, remainder_v);
 
-        let (dividend, inter_mul_value) = divisor.mul_add(&outcome, &r, lkm, true);
         config
-            .inter_mul_value
-            .assign_mul_outcome(instance, lkm, &inter_mul_value)?;
+            .is_divisor_zero
+            .assign_instance(instance, (divisor as u64).into())?;
+
+        config.is_remainder_lt_divisor.assign_instance(
+            instance,
+            lkm,
+            rem_pos as u64,
+            div_pos as u64,
+        )?;
 
-        config.dividend.assign_add_outcome(instance, &dividend);
-        config.remainder.assign_limbs(instance, r.as_u16_limbs());
-        config
-            .is_zero
-            .assign_instance(instance, divisor.as_u64().into())?;
-        config
-            .remainder_lt
-            .assign_instance(instance, lkm, r.as_u64(), divisor.as_u64())?;
+        config.r_insn.assign_instance(instance, lkm, step)?;
 
         Ok(())
     }
 }
 
+// TODO Tests
```

Comment:
I guess you forgot this `TODO`. I didn't see any testing added.

Reply:
Not forgotten!  I mentioned that this is still TODO in the PR description -- another member of our team is working on it, but paused yesterday due to the rapid changes from the review.  Should be finished relatively soon, but probably an initial review of the circuit design would still be valuable at this point before we validate with unit tests.

Reply:
I tried to write a simple test, but it kept failing.  (Not sure that's because I made some mistakes, or because the code here is wrong.)

So I would suggest that we only merge this PR once we have some passing tests.